Allow for statefulSet Teleport-cluster Deployments

yjperez opened the pull request:
Created a resource mode to allow the Auth to be deployed in a StatefulSet so that when deployed with persistentVolumes it could be deployed on separate nodes.

yjperez committed:
Creating an Auth-Statefulset template yaml file

yjperez committed:
Adding IF statement for resource type

yjperez committed:
Adding resourceType to the values schema

yjperez committed:
Modifying Values file to account for statefulset

hugoShaka commented on examples/chart/teleport-cluster/templates/auth/statefulset.yaml:
```diff
@@ -0,0 +1,278 @@
+      - name: "config"
+        configMap:
+          name: {{ .Release.Name }}-auth
+      - name: "data"
+        {{- if and ($auth.persistence.enabled) ( and (not (eq $auth.chartMode "gcp")) (not (eq $auth.chartMode "aws"))) }}
+        persistentVolumeClaim:
+          claimName: {{ if $auth.persistence.existingClaimName }}{{ $auth.persistence.existingClaimName }}{{ else }}{{ .Release.Name }}{{ end }}
+        {{- else }}
```
If I understand correctly, the whole PR goal is to be able to mount per-pod volume to have a persistent buffer before audit log and event ingestion. In the current state, the statefulset does not currently solve this issue as it still mounts a single PVC on all pods.

AFAIU we need to remove the single PVC mount (and the single PVC) and shift to a volume template in the statefulset. We also need to explain that when switching this value you loose your existing PVC. This is not a bad thing as it will block people from shooting themselves in the foot by enabling statefulset without understanding what data is stored where.

Note: PVC volumes cannot be resized, this must be mentioned when we'll expose the PVC volume settings like disk type and size.

webvictim replied:
Now uses a volume template

yjperez commented on examples/chart/teleport-cluster/values.yaml:
```diff
@@ -252,6 +252,11 @@ labels: {}
+# Deploy Teleport using a Deployment or StatefulSet. The default is "deployment". Options:
+# - "deployment": will deploy both the auth and proxy using a deployment.
+# - "statefulset" will deploy both the auth as a statefulSet and proxy as a deployment
+resourceType: statefulset 
```
You are correct, oversight on my side. I thought I had changed it back from my testing.

yjperez committed:
Update values.yaml

Changing default resourceType to deployment

yjperez committed:
Update examples/chart/teleport-cluster/templates/auth/statefulset.yaml

Co-authored-by: Hugo Shaka <hugo.hervieux@goteleport.com>

webvictim commented:
I'll help write the tests as agreed @yjperez - just wanted to get Hugo's initial review first.

webvictim committed:
Added unit tests

A few are currently broken, will fix the underlying problems in another commit

webvictim commented on examples/chart/teleport-cluster/templates/auth/statefulset.yaml:
Added tests in be145f8.

Some are broken due to differences in volume mounts etc, will take a look.

webvictim replied:
Unit tests should be fixed in 52f8f34

webvictim replied:
Unit tests should be fixed in 52f8f34

webvictim committed:
Add missing space

webvictim commented on examples/chart/teleport-cluster/values.yaml:
```diff
@@ -426,6 +431,21 @@ tls:
+  whenDeleted: Retain
+  whenScaled: Retain
+
+# set the updatestrategy to RollingUpdate or OnDelete
+updateStrategy:
+  type: RollingUpdate
```
I personally believe we should be very proscriptive about the way the chart is used when `deploymentType` is `statefulset`, as it's non-standard and has the potential to introduce non-deterministic behaviour.

I'm fine with removing the ability to set `updateStrategy` and assuming that `RollingUpdate` will always be used. If people want different behaviour, they can open a feature request or fork the chart themselves and add the functionality.

webvictim replied:
I removed `updateStrategy`

webvictim commented on examples/chart/teleport-cluster/values.yaml:
```diff
@@ -252,6 +252,11 @@ labels: {}
 # (3) When set to "scratch", you must write the teleport configuration in auth.teleportConfig and proxy.teleportConfig.
 chartMode: standalone
 
+# Deploy Teleport using a Deployment or StatefulSet. The default is "deployment". Options:
```
Added some details.

webvictim committed:
Added volumeClaimTemplates to statefulset

Added support for storageClassName

Added/removed unit tests as apporpriate

Fixed up documentation in values file

Removed support for updateStrategy as changing it will complicate deployments

webvictim committed:
Merge branch 'master' into jay/helm-cluster-statefulset

webvictim committed:
Fix line break

webvictim committed:
Update test snapshots to use distroless images

webvictim commented:
Two main remaining things to do on this PR:
- update docs in the `teleport-cluster` chart reference to include `resourceType`, `persistence.storageClass` and a description of when you might want to use each
- actually test this in a cluster to make sure a separate PVC is successfully created for each pod, storage is separated and persists across pod restarts

yjperez commented:
> 

Thanks @webvictim for taking lead on these changes this week, haven't had the cycles this week to work on this. I will test the PVC with stateful set
